### ISSUE TITLE
publish() method have problems with pushing changes, when SSH key is used?

### ISSUE THREAD

**Issue by oskar-j** [ramnathv/slidify#419]
I finally managed to configured little my .git settings for my slidify project.
But because I have 2-key authentication on my GitHub profile, I'm obliged to use SSH putty keys with my git scm / tortoise git. 
I had couple of different errors, one of them I fixed with "git config" as described in issue 322 (without using the --global option!), but this is the last error I cannot fix in any way:

```r
> publish(user = "oskar-j", repo = "slidify-presentations-topic1")
Publishing deck to oskar-j/slidify-presentations-topic1
warning: You ran 'git add' with neither '-A (--all)' or '--ignore-removal',
whose behaviour will change in Git 2.0 with respect to paths you removed.
Paths like '.cache/unnamed-chunk-10_829cd7efc87b7616c51f75453d3cc18c.RData' that are
removed from your working tree are ignored with this version of Git.

* 'git add --ignore-removal <pathspec>', which is the current default,
  ignores paths you removed from your working tree.

* 'git add --all <pathspec>' will let you also record the removals.

Run 'git status' to check the paths you removed from your working tree.

[gh-pages 0c19949] publishing deck
 9 files changed, 526 insertions(+), 10 deletions(-)
 create mode 100644 .Rhistory
 create mode 100644 .cache/unnamed-chunk-10_1c0400872d6b7949c53acb5de0e98873.RData
 rename .cache/{unnamed-chunk-10_829cd7efc87b7616c51f75453d3cc18c.rdb => unnamed-chunk-10_1c0400872d6b7949c53acb5de0e98873.rdb} (100%)
 rename .cache/{unnamed-chunk-10_829cd7efc87b7616c51f75453d3cc18c.rdx => unnamed-chunk-10_1c0400872d6b7949c53acb5de0e98873.rdx} (100%)
 delete mode 100644 .cache/unnamed-chunk-10_829cd7efc87b7616c51f75453d3cc18c.RData
 create mode 100644 sentiment-through-time.xlsx
```

R Console freezes. Always. I press "stop".

Than I get:

```r
Warning message:
running command 'git push git@github.com:oskar-j/slidify-presentations-topic1 gh-pages' had status 8240 
```

And the only way to push my changes is to use a seperate windows console and lunch the instruction shown up `git push git@github.com:oskar-j/slidify-presentations-topic1 gh-pages`

Thank you for your great support!
B.R.,
Oskar.

**Comment by oskar-j** [comment]
Moreover, when I create a new slidify project, I get this warning and error messages:

```r
> author("github-users-gender")
Creating slide directory at github-users-gender...
Copying files to github-users-gender...
Finished creating slide directory...
Switching to slide directory...
Initializing Git Repo
Initialized empty Git repository in C:/big data/github-gender-slidify/github-users-gender/.git/
error: pathspec 'Commit'' did not match any file(s) known to git.
Checking out gh-pages branch...
Switched to a new branch 'gh-pages'
Adding .nojekyll to repo
Opening slide deck...
Warning message:
running command 'git commit --allow-empty -m 'Initial Commit'' had status 1 
```

Because @ramnathv seems to be busy with side projects, is there any git hacker which can help me to set properly git-scm settings for slidify?

Many thanks!
Oskar.

**Comment by ptressel** [comment]
Hi, Oskar!  I ran into the same git commit failure issue with author() on Windows 7.  The problem is the use of single quotes around 'Initial Commit'.  As you can see from the error message, that string was treated as two separate tokens: only 'Initial was treated as the argument to -m, and the Commit' was regarded as a pathspec by git commit.

You can *temporarily* fix this by editing the slidify:::init_repo function.  Do this to edit that function:

fixInNamespace("init_repo", "slidify")

Not sure what that does on the R command line, but in RStudio (and, one presumes, RGui) it brings up an editor with the function.  Then change the single quotes ' to escaped double quotes \" like this:

system("git commit --allow-empty -m \"Initial Commit\"")

Note that edit will only last for the current session -- when slidify is loaded again, it will not have that edit.

I don't have access to a Linux or MacOS system to test that on, but believe it should be fine there too.  If I can get someone to test it, I'll send a PR.

About git config:  R sets an environment variable HOME, and on Windows, sets that to the user's Documents folder, which is incorrect -- that is not a Windows user's home directory, and it is not what git expects.  The proper Windows user directory is the user profile folder, typically C:\Users\Name-of-user.  That is available as the Windows environment variable %USERPROFILE%.  (Note, to be truly proper, configuration data should go in the AppData folder, which is inside the user profile folder.  But many ported Linux / Unix apps put config files directly in the user profile folder -- this works too.)

Anyhow, what's happening is that R sets HOME to the "wrong" place, Documents, and poor git, when run from R, sees that %HOME% and tries to find .gitconfig there, and fails, because when git installed itself, it put .gitconfig right in %USERPROFILE%.

I wanted git to work properly from R, so had to change what R used for HOME.  One can do this by setting either R_USER or HOME as a Windows user (not system) environment variable -- set it to %USERPROFILE%.  I set R_USER because I didn't want to set something generic like HOME that might affect other programs.  The right time to do this is before installing R, because R will create an r folder there, and put win-library inside it.  Since I was fixing this up after installing R, I moved the r folder out of Documents to my C:\Users\My-name folder.  This also changes the meaning of "~".  So I had to close and re-open all the files open in the editor in RStudio, as it was using relative paths.  I have assorted notes and scripts with setwd("~/blah/blah") in them, so those will need changing as they come up.  Again, it's better to do this before installing R.  Yet better would be for R itself to use the real Windows user folder for HOME, same as other Linux apps, not the Documents folder...

Note when you omit --global from git config, you are setting options for the current repo's .git/config file, not in your .gitconfig file.  This is equivalent to git config --local, which is the default.

**Comment by ptressel** [comment]
Just FYA, here is someone running into the identical single quote problem with mercurial under windows:

http://superuser.com/questions/324278/how-to-make-windows-command-prompt-treat-single-quote-as-though-it-is-a-double-q

**Comment by ramnathv** [comment]
Thanks @ptressel for diagnosing the root cause. Would switching to double quotes solve the problem and ensure that it still works across platforms? If you can do some testing and send me a PR I would be happy to merge this change in.

**Comment by ptressel** [comment]
It works on Windows, and double quotes should be fine on Linux / Unix.  I've certainly used double quotes around commit messages on Linux.

I put out a call to fellow students in a data science MOOC to try this under MacOS and Linux.  If I don't get a response soon, I'll try it on Linux in a VM, and try to talk friends with Macs into trying it.  :D

(The R HOME versus git config issue is, of course, nothing to do with slidify.  The only thing one might do is include a note in docs about what's happening.  Better if there were an external post about it to link to.)

**Comment by ptressel** [comment]
It's been tried under Linux, and is ok there.  Still looking for a Mac tester.  ;-)

**Comment by dalupus** [comment]
I can test on a mac for you.  What exactly would you like to try?

**Comment by ptressel** [comment]
Great, dalupus, and thank you!

First, either modify the git commit line as above, or (re)install slidify from https://github.com/ptressel/slidify using devtools::install_github("ptressel/slidify").  (The latter is better as it will check the actual change.)  Then try making a fresh presentation with:

author("some_directory")

Hoped outcome is that there will be no change from before -- both types of quotes should be ok on a Mac.

**Comment by dalupus** [comment]
No issues on mac

~~~~
> author("testSlides4")
Creating slide directory at testSlides4...
Copying files to testSlides4...
Finished creating slide directory...
Switching to slide directory...
Initializing Git Repo
Initialized empty Git repository in C:/Users/dalupus/Documents/FAU/PerfModeling/testSlides3/testSlides4/.git/
[master (root-commit) cefb713] Initial Commit
Checking out gh-pages branch...
Switched to a new branch 'gh-pages'
Adding .nojekyll to repo
Opening slide deck...
~~~~

**Comment by ptressel** [comment]
Thanks, dalupus!!  I'll send a PR.

**Comment by ptressel** [comment]
Of course, that wasn't oskar-j's original issue.  Sounds like there might be a git version issue.  I'm still at git 1.9.5, so maybe I could reproduce it.  I too use SSH keys with Github.  But if it's a problem with git not finding its global config, that might be harder for me to repro, as I already changed R_USER and moved my R config.

So, a question for oskar-j:  What version of git are you using?

**Comment by ptressel** [comment]
Some info about those warnings -- apparently the behavior of git changed in 2.x, and the command that's being done will not behave the same way under 2.x.

http://stackoverflow.com/questions/23027678/warning-you-ran-git-add-with-neither-a-all-or-ignore-removal
https://reprog.wordpress.com/2014/08/29/dear-git-stop-changing-default-behaviour/

So it might be that there is a git add command where what's needed is to pick the desired behavior (include deleted files in the git add so they get deleted from the repo, or ignore deleted files and don't delete them), and add the appropriate flag.  If that's what's happening, then I should be able to repro it.
